### PR TITLE
Implement is_substring_palindrome using dynamic programming.

### PR DESCRIPTION
**Pull request by Erfaniaa**
It can be in strings or dynamic programming directories, both are OK.

### CODE REVIEW

File: strings/is_substring_palindrome.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+text = ""
+dp = [[False for i in range(MAX_LEN)] for i in range(MAX_LEN)]
+
+def preprocess(s):
+    """
+        Preprocesses a string using dynamic programming.
+        Time complexity: O(n^2)
+    """
+    global text
```
Comment (SergeyTsaplin) [inline strings/is_substring_palindrome.py]: don't use global vavriables here. It'll be much better if you return the matrix (`dp`) as a result, the `text` variable is useless at all.

```py
def process(s):
    ...
    n = len(s)
    matrix = [[False for i in range(n)] for i in range(n)]
    ...
    
    return matrix

...
if __name__ == '__main__':
    ...
    matrix = process(s)
    ...
    if is_polindrome(matrix, l, r):
        ...
```

So you can delete the globals `MAX_LEN`, `text` and `dp`

File: strings/is_substring_palindrome.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
+MAX_LEN = 1000
```
Comment (SergeyTsaplin) [inline strings/is_substring_palindrome.py]: Module-level docstring with the algorithm description should be here

File: strings/is_substring_palindrome.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    global text
+    global dp
+    text = s
+    n = len(s)
+    dp = [[False for i in range(n)] for i in range(n)]
+
+    for i in range(n):
+        dp[i][i] = True
+
+    for i in range(n - 1):
```
Comment (SergeyTsaplin) [inline strings/is_substring_palindrome.py]: Seems like the iteration can be avoided. Move this to the previous one:

```py
for i in range(n):
    dp[i][i] = True
    if i < n - 1:
        dp[i][i + 1] = (text[i] == text[i + 1])  
```

File: strings/is_substring_palindrome.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        dp[i][i + 1] = (text[i] == text[i + 1]) 
+
+    for substr_len in range(2, n + 1):
```
Comment (SergeyTsaplin) [inline strings/is_substring_palindrome.py]: It's possible to avoid the additional iteration

File: strings/is_substring_palindrome.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
+MAX_LEN = 1000
+
+text = ""
+dp = [[False for i in range(MAX_LEN)] for i in range(MAX_LEN)]
+
+def preprocess(s):
+    """
```
Comment (SergeyTsaplin) [inline strings/is_substring_palindrome.py]: Very strange format of the docstring. Also it's better to add a description what the function exactly does

File: strings/is_substring_palindrome.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        dp[i][i + 1] = (text[i] == text[i + 1]) 
+
+    for substr_len in range(2, n + 1):
+        for i in range(n - substr_len + 1):
+            j = i + substr_len - 1
+            dp[i][j] = (text[i] == text[j] and dp[i + 1][j - 1])
+
+
+def is_substring_palindrome(l, r):
+    """
```
Comment (SergeyTsaplin) [inline strings/is_substring_palindrome.py]: Strange format of the docstring

**Comment by varunu28** [comment]
Cannot add a review as this is an interviewing problem and not a specific computer science algorithm. Not sure how does it works in the Python project of "The Algorithms" but I have always advocated that the repo should have efficient implementations for well-known algorithms and not solutions to problems of online judges.

File: strings/is_substring_palindrome.py (hunk 6 above)
Comment (Erfaniaa) [reply strings/is_substring_palindrome.py]: I updated the code. It's now OK. Please review it again.

**Comment by Erfaniaa** [comment]
This code is a classic usage of dynamic programming to detect palindromes. It's popular and well-known (like minimum edit distance problem), but it doesn't have a special name or label (like the knapsack, KMP, Dijkstra, etc). Not having a special doesn't mean that it's non-classical or not well-known.
This shows how to use dynamic programming to answer queries. It also shows a good comparison between an easy solution and Manacher's algorithm.
So, I hope it can help people who are reading the string and dynamic programming algorithms in this repository to learn something.
@varunu28

**Comment by varunu28** [comment]
@Erfaniaa That is exactly what my point. I do not think adding all possible variations or implementations of an algorithm is necessary for "The Algorithms" repo. Those variations are the next steps for anyone trying to learn algorithms once they are done with basics. Dynamic Programming has few paradigms such as bottom up or top down. If they are already covered in the repo then it does not make sense to add more variations of the same. 

We should be rather focussing on adding algorithms from other areas of computer science such as distributed systems, operating systems or NLP rather than adding solutions to interview problems in the repo.

**Comment by ParthS007** [comment]
I totally agree with @varunu28 :+1: This is out of the scope of this repository.

**Pull request closed without merging**